jolt-core/src/subprotocols/toom.rs

```diff
@@ -0,0 +1,911 @@
+    let p_x0 = p[1] - p[0]; // x-slope at y = 0
+    let p_x1 = p[3] - p[2]; // x-slope at y = 1
+    let p_y0 = p[2] - p[0]; // y-slope at x = 0
+    let p_y1 = p[3] - p[1]; // y-slope at x = 1
+    let p_xy = p_x1 - p_x0; // mixed term = p11
+
+    // ---- same for q ----
+    let q_x0 = q[1] - q[0];
+    let q_x1 = q[3] - q[2];
+    let q_y0 = q[2] - q[0];
+    let q_y1 = q[3] - q[1];
+    let q_xy = q_x1 - q_x0;
+
+    // ---- point-wise products (9 mults) ----
+    [
+        // y = 0 row
+        p[0] * q[0], // (0,0)
+        p[1] * q[1], // (1,0)
+        p_x0 * q_x0, // (∞,0)
+        // y = 1 row
+        p[2] * q[2], // (0,1)
+        p[3] * q[3], // (1,1)
+        p_x1 * q_x1, // (∞,1)
+        // y = ∞ row
+        p_y0 * q_y0, // (0,∞)
+        p_y1 * q_y1, // (1,∞)
+        p_xy * q_xy, // (∞,∞)
+    ]
+}
+
+// -----------------------------------------------------------------------------
+//  d = 3, n = 2  (three multilinear polynomials in 2 variables)
+//  • inputs:  three arrays length 4 each with order (00,10,01,11)
+//  • outputs: product on {0,1,−1,∞}² (16 points), row-major x-order [0,1,−1,∞]
+// -----------------------------------------------------------------------------
+#[inline]
+pub fn eval_toom3_2<F: FieldMulSmall>(polys: [[F; 4]; 3]) -> [F; 16] {
+    // Step 1: quadratic grid from first two polys
+    let a = eval_toom2_2(polys[0], polys[1]);
+    let idx3 = |x: usize, y: usize| -> usize { y * 3 + x };
+    // Build 4×4 grid for A
+    let mut a_grid = [[F::from_u64(0); 4]; 4];
+    // helper closure for x-extension
+    #[inline]
+    fn quad_x_m1<F: FieldMulSmall>(a0: F, a1: F, a_inf: F) -> F {
+        dbl(a0) - a1 + dbl(a_inf)
+    }
+    // rows y=0,1,∞
+    for (src_y, dst_y) in [(0usize, 0usize), (1, 1), (2, 3)] {
+        let a0 = a[idx3(0, src_y)];
+        let a1 = a[idx3(1, src_y)];
+        let a_inf = a[idx3(2, src_y)];
+        let a_m1 = quad_x_m1(a0, a1, a_inf);
+        a_grid[dst_y] = [a0, a1, a_m1, a_inf];
+    }
+    // y = −1 row (dst 2)
+    for x in 0..4 {
+        let a0 = a_grid[0][x];
+        let a1 = a_grid[1][x];
+        let a_inf = a_grid[3][x];
+        a_grid[2][x] = quad_x_m1(a0, a1, a_inf);
+    }
+
+    // Step 2: third linear poly values on grid
+    let p = polys[2];
+    let a0 = p[0];
+    let bx = p[1] - p[0];
+    let by = p[2] - p[0];
+    let bxy = p[3] - p[1] - p[2] + p[0];
+    let xs = [Some(F::zero()), Some(F::one()), Some(-F::one()), None];
+    let ys = xs;
+    let mut b_grid = [[F::zero(); 4]; 4];
+    for (i_y, y_opt) in ys.iter().enumerate() {
+        for (i_x, x_opt) in xs.iter().enumerate() {
+            b_grid[i_y][i_x] = match (x_opt, y_opt) {
+                (Some(xv), Some(yv)) => a0 + bx * (*xv) + by * (*yv) + bxy * (*xv) * (*yv),
+                (None, Some(yv)) => bx + bxy * (*yv),
+                (Some(xv), None) => by + bxy * (*xv),
+                (None, None) => bxy,
+            };
+        }
+    }
```

Comment:
idk if we'll end up using this function but we can avoid some multiplications by unrolling this loop and handling each of the 16 cases of (x, y) individually